### PR TITLE
API can store and edit sections

### PR DESCRIPTION
**Pull request by tochman**
PT: https://www.pivotaltracker.com/story/show/179419149

### CODE REVIEW

**Review by SevaDer14 — COMMENTED** [review]
Two big things to consider:
1) Do we need create functionality for sections in this iteration?
2) Do we need View model or it is easier to just have it as an attribute of section?
I say No to both questions, if no objections we just need to make seeds for sections and buttons, refactor so tests pass and we are almost there.

File: app/models/section.rb [hunk 1]
```diff
@@ -1,5 +1,9 @@
 class Section < ApplicationRecord
-  enum variant: { services: 0, about_us: 1, about_self_care: 2, information: 3 }
+  enum variant: { regular: 0, no_image: 1, carousel: 2, slider: 3 }
   belongs_to :view
   validates_presence_of :header
+  has_one :image, ensuring_owner_was: regular || slider || carousel, dependent: :destroy_async
+  has_many :buttons, dependent: :destroy, class_name: 'Cta'
+  validates_associated :buttons
+  has_many :slides, ensuring_owner_was: slider, dependent: :destroy, class_name: 'Image'
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply app/models/section.rb:8]: removed

File: app/models/cta.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+class Cta < ApplicationRecord
+  belongs_to :section
+  validate :is_regular?
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply app/models/cta.rb:4]: removed

File: spec/factories/sections.rb [hunk 3]
```diff
@@ -3,6 +3,23 @@
...
     header { "MyString" }
     description { "MyText" }
-    variant { 1 }
+    factory :regular do
+      variant { :regular }
+    end
+
+    factory :slider do
+      variant { :slider }
+    end
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply spec/factories/sections.rb:12]: done

File: spec/models/section_spec.rb [hunk 4]
```diff
@@ -1,7 +1,43 @@
...
+        expect(subject.variant_before_type_cast).to eq 2
+      end
+    end
+
+    context 'slider section' do
+      subject { create(:slider) }
+      it { is_expected.to be_valid }
+      it 'is expected to store 3 as numeric value for variant' do
+        expect(subject.variant_before_type_cast).to eq 3
+      end
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply spec/models/section_spec.rb:40]: done

File: app/models/view.rb [hunk 5]
```diff
@@ -1,5 +1,7 @@
 class View < ApplicationRecord
-
   validates_presence_of :name
-  has_many :sections
+
+  enum variant: { services: 0, about_us: 1, about_self_care: 2, information: 3 }
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply app/models/view.rb:4]: it works so no reason to change it

File: db/migrate/20210904054745_add_variant_to_views.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddVariantToViews < ActiveRecord::Migration[6.1]
+  def change
+    add_column :views, :variant, :integer
+  end
+end
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply db/migrate/20210904054745_add_variant_to_views.rb:5]: it works so no reason to change it

File: app/serializers/section/show_serializer.rb [hunk 7]
```diff
@@ -1,3 +1,9 @@
 class Section::ShowSerializer < ActiveModel::Serializer
-  attributes :variant, :header, :description 
+  attributes :variant, :header, :description, :view_id 
+
+  def attributes(*args)
+    hash = super
+    hash[:buttons] = object.buttons if object.regular?
+    hash
+  end
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply app/serializers/section/show_serializer.rb]: done

File: spec/models/view_spec.rb [hunk 8]
```diff
@@ -10,10 +10,15 @@
   end
 
   describe 'Associations' do
-    it { is_expected.to have_many(:sections) }
+    it { is_expected.to have_many(:sections).dependent(:destroy) }
   end
 
   describe 'Database' do
     it { is_expected.to have_db_column(:name).of_type(:string) }
+    it { is_expected.to have_db_column(:variant).of_type(:integer) }
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply spec/models/view_spec.rb:18]: it works so no reason to change it

File: spec/models/view_spec.rb [hunk 9]
```diff
@@ -10,10 +10,15 @@
...
+  end
+
+  describe 'Enums' do
+    it { is_expected.to define_enum_for(:variant).with_values(%i[services about_us about_self_care information]) }
   end
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply spec/models/view_spec.rb:23]: it works so no reason to change it

File: app/controllers/sections_controller.rb [hunk 10]
```diff
@@ -1,8 +1,24 @@
...
+  end
+
+  def create
+    section = Section.create section_params
+    render json: section, serializer: Section::ShowSerializer, status: :created
+  end
+
+  def section_params
+    params.require(:section).permit(:view_id, :header, :description)
+  end
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply app/controllers/sections_controller.rb:23]: it works so no reason to change it

File: spec/requests/api/cards/update_spec.rb [hunk 11]
```diff
@@ -0,0 +1,100 @@
+RSpec.describe 'PUT /api/articles/:id' do
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply spec/requests/api/cards/update_spec.rb:1]: true, fixed

File: spec/models/cta_spec.rb [hunk 12]
```diff
@@ -0,0 +1,12 @@
+RSpec.describe Cta, type: :model do
+  describe 'Factory' do
+    it 'is expected to have valid basic Factory' do
+      expect(create(:button)).to be_valid
+    end
+  end
+  describe 'db table' do
+    it { is_expected.to have_db_column(:text).of_type(:string) }
+    it { is_expected.to have_db_column(:link).of_type(:string) }
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply spec/models/cta_spec.rb:9]: done

File: db/migrate/20210904061251_create_cta.rb [hunk 13]
```diff
@@ -0,0 +1,11 @@
+class CreateCta < ActiveRecord::Migration[6.1]
+  def change
+    create_table :cta do |t|
+      t.string :text
+      t.string :link
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply db/migrate/20210904061251_create_cta.rb:5]: changed

File: spec/factories/sections.rb [hunk 14]
```diff
@@ -3,6 +3,19 @@
...
+
+    factory :carousel do
+      variant { :carousel }
+    end
   end
 end
+
+
+# regular: 0, no_image: 1, carousel: 2
```
[the remark this answers is not in the document]
Comment (SevaDer14) [reply spec/factories/sections.rb:21]: Probably forgotten, removed

**Review by 1CIM — APPROVED** [review]
LGTM

**Pull request merged**